### PR TITLE
feat(tokenless): reduce large JSON records

### PR DESCRIPTION
**Pull request by boluochoufeng**
## Why

Large JSON object arrays currently fall through to positional truncation, which can discard failures and anomalies in the middle of a tool result. This change adds deterministic, recoverable record selection while preserving the PostTool architecture's single JSON compressor and single final arbitration point.

## What changed

`JsonCompressor` now evaluates complete-data compact JSON or TOON first; savings of at least 15% immediately select that candidate and skip bounded work, while smaller savings continue to bounded generation and final arbitration. Recoverable arrays of at least 33 objects use a 32-record base budget that preserves boundaries, error signals, shape anomalies, numeric outliers, and stable samples; critical records may exceed the base budget, and the complete original array is stored once for retrieval. CLI dry-runs validate the configured recovery path but use an in-memory Stash for candidate measurement. The new operation is exposed through Rust/Python protocol DTOs, stats, tests, benchmarks, and user documentation without adding a compressor or plugin abstraction.

## Related issue

no-issue: incremental PostTool compression architecture work

## User / Agent impact

Large JSON record arrays can now achieve substantially higher savings while retaining important records and a marker for exact recovery. Adapters without a trusted recovery path remain lossless-only and do not apply record reduction.

## Risk and compatibility

- [x] Public CLI, API, configuration, or documented behavior changed
- [ ] Privileged or security-sensitive behavior changed
- [x] Cross-component contract changed
- [ ] Migration or rollback guidance is needed

The response operation enum gains `json_record_reduction`; request shapes and configuration remain unchanged. The behavior is limited to JSON object arrays with at least 33 entries when recovery is available, and can be rolled back by reverting this PR.

## Validation

- `cargo fmt --all -- --check`
- `cargo clippy --workspace --all-targets --locked -- -D warnings`
- `cargo test --workspace --locked`
- `cargo doc --workspace --no-deps`
- `make test` with the locally built Tokenless and RTK binaries on `PATH`
- `make test-python-runtime` with Miniforge Python providing `venv`
- L1 benchmark workspace: `RUST_MIN_STACK=16777216 cargo test --locked`, plus fmt and clippy
- L2 benchmark workspace: `cargo test --locked`, plus fmt and clippy

Coverage includes the 15% boundary, base budget, critical-record overflow, nested arrays, stable ordering, JSON/TOON selection, Stash failure, authorized byte-exact recovery, CLI dry-run without persistent Stash entries, timeout, and single commit/rollback behavior.

## Documentation and rollback

README files and the response-compression and Stash guides describe the new selection and recovery behavior. No changelog entry is included; rollback is a direct revert of this PR.

### CODE REVIEW

**Review by ikunkun-sys — COMMENTED** [review]
Found one non-blocking documentation/privacy contract issue on the reviewed head. I did not repeat the existing omitted-count P2 thread. No P0/P1 findings.

File: src/tokenless/crates/tokenless-compressors/src/json.rs [hunk 1]
```diff
@@ -379,6 +610,15 @@ impl<'a> Session<'a> {
     }
 
     fn compress_array(&mut self, values: &[Value], depth: usize) -> Value {
+        if self.bounds_enabled && is_record_array(values) {
```
Comment (ikunkun-sys) [inline src/tokenless/crates/tokenless-compressors/src/json.rs:613]: **[P2] Canonical guides still describe the old array and Stash contract**

This branch sends every array of at least 33 objects through the fixed 32-record reduction path, and `reduce_records` stashes the complete array. The canonical EN/ZH CLI references still say `--truncate-arrays-at` controls array retention and that Stash contains only the dropped middle segment; the privacy guides list the same old storage boundary. On this head, 40 object records with `--truncate-arrays-at 1 --array-tail-preserve 0` emit 32 records plus a marker, while `--no-stash` keeps all 40. Users therefore cannot infer the CLI behavior or the full sensitive-data footprint of `stash.db` from the canonical user guides.

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

架构上干净地实现了"先无损、后有界"的两级候选选择：Record Reduction 保持了单一压缩器、单一仲裁点和单次 Stash 事务，选择逻辑确定性好、测试覆盖充分（含边界、异常保留、Stash 失败、dry-run、回滚）。本地验证：`cargo fmt` / `clippy -D warnings` 通过，主工作区 705 测试、L1 97 测试、L2 42 测试全部通过。以下为增量发现（已有 Codex review 指出的 `reduce_records` omitted 计数问题不再重复）。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

- 无

### 🟡 建议修改（不阻塞但推荐）

- [crates/tokenless-cli/src/main.rs:511-521] CLI dry-run 无法预览 Record Reduction，预览保真度回退。`needs_stash` 仍要求 `compression_on`，dry-run 时 CLI 不打开 Stash store，因此 `pipeline.rs:88-101` 新增的 `dry_run_store`（InMemoryStore）机制经 CLI 路径不可达（只有 SDK/Python runtime 这种无条件传 store 的调用方受益）。实测：同一 200 条记录数组，`TOKENLESS_COMPRESSION_ENABLED=1` 时 6096→999 tokens（-83.6%，`json_record_reduction`）；`TOKENLESS_COMPRESSION_ENABLED=0` dry-run 报 `no_savings`。改动前该场景 dry-run 还能预览位置型截断的节省，现在为 0，会误导 README 推荐的 0/1 双跑对照。既然 `dry_run_store` 是临时 store、不会孤儿化持久条目，建议放开 CLI dry-run 的 stash 挂载（或至少在文档中说明该限制）。

- [crates/tokenless-compressors/src/json.rs:180-201] 低于 15% 的无损候选仍可能替换模型内容，TOON 场景收益趋近于 0。15% 门槛只决定是否跳过 bounded 会话，不约束最终接受：当 bounded 候选无效或不更小时，哪怕无损候选只省极少也会被仲裁器接受。实测 cosh-ng 契约 golden（tests/contract/goldens/post_tool/bash_array_truncation/cosh-ng.json）即此情形：200 条字符串数组被编码为 TOON，1732→1731 tokens（约 0.06%），模型内容格式被改变但几乎无收益。虽然无损且属文档描述的算法行为，但建议对最终接受也施加有意义的节省门槛（至少对 TOON 表示切换），或在文档中明确"低于 15% 的无损候选在无可比有界候选时仍会被采用"。

- [crates/tokenless-compressors/src/json.rs:166-201] 未过 15% 门槛的 payload 现在走两遍完整压缩（full 会话 + bounded 会话），大输入（接近 `max_input_bytes`）约 2 倍 CPU，且 pipeline 超时（`RESPONSE_PIPELINE_TIMEOUT`）是在 compress 返回后才判定，超时会直接 passthrough。建议关注 L1/L2 延迟指标，必要时对超大输入加快速路径。

- [crates/tokenless-compressors/src/json.rs:435-453]（`has_error_signal`）错误信号启发式只覆盖字符串型状态字段：数值状态码（如 `"status": 500`）和布尔失败标志不会命中 `STATUS_FIELDS` 分支（数值 500 可能经 2σ 离群值路径部分兜底）。不阻塞，可后续迭代补充常见数值/布尔形态。

### 🟢 值得肯定

- 选择逻辑完全确定（首尾 + 错误信号 + 结构异常 + 2σ 离群 + 去重后稳定等距采样），采样位置公式无越界风险，且有专门的稳定性与预算溢出测试。
- 每个缩减数组只写一条 Stash（完整原数组），配合 `StashLedger` 的 commit/rollback，被拒绝候选的临时写入会被正确删除（`record_candidate_with_no_savings_rolls_back_its_single_write` 等测试验证）；无可用 Stash 时明确不做 reduction、也不退化为位置型截断，符合"无可信恢复路径保持无损"的设计。
- marker key 为 24 位十六进制 BLAKE3，`tokenless retrieve '<key>'` 无注入风险。
- 文档（response-compression.md / stash-reversible-compression.md / README）与 benchmark 基线、契约 golden 均如实重定基，未掩盖指标变化。

**Review by Forrest-ly — COMMENTED** [review]
## 详细意见（非阻塞）

### 1. 无损候选在低于 15% 收益（甚至 1 token）时仍可被采用
`crates/tokenless-compressors/src/json.rs:180-201`

15% 门槛（第 180 行 `saves_at_least_percent`）只控制"跳过 bounded 尝试"的提前返回；当 bounded 候选不合法时（例如 Common Hook：`retrieval_available=false` + `require_reversibility`，bounded 截断候选为 Unrecoverable），第 195 行的选择逻辑会回退到无损候选——只要它带有任何操作，哪怕只省 1 个 token。

实测（bash_array_truncation fixture，6926 字符，1732→1731 tokens，约 0.06%）：`cosh-ng`（`replace_with_text=true`）现在会 applied 一个 TOON 表示，而父提交 `6712b3d6` 上是 `recoverability_unavailable` 透传。该变化无损、golden 也是有意重新基线化的，但 PR 描述"prefers complete-data compact JSON or TOON **when it saves at least 15%**"与实际行为不完全一致（`docs/response-compression.md` 的中文描述"否则再生成 Bounded 候选，并选择更小的合法表示"是准确的）。建议：对齐 PR 描述措辞；或考虑对"改变表示形式"（TOON）的候选设置一个最小收益下限（若干 token），避免为 1 token 的收益切换模型可见的表示。

### 2. `records_omitted` 在极端情况下会少计
`crates/tokenless-compressors/src/json.rs:673-695`

`omitted = values.len() - selected.len()` 在 `push_if_kept`（第 684 行）递归压缩被选记录**之前**计算。若 `drop_nulls`/`drop_empty_fields` 开启，个别被选记录压缩后可能变为 `{}` 而被过滤掉，此时输出里可见的记录数少于 `selected.len()`，Marker "<... N of M records omitted>" 与 `metrics.records_omitted` 都会少计。无数据丢失（Stash 中保留完整原数组）。建议把计数改为基于最终输出长度，或在注释中明确这一口径。

### 3. `has_error_signal` 的状态子串匹配存在误报方向
`crates/tokenless-compressors/src/json.rs:435-457`

`STATUS_SIGNALS` 用 `status.contains(signal)`，因此 `"0 errors"`、`"no timeout observed"` 这类文本也会命中，使记录被保留。误报方向是安全的（多保留、且可恢复）且行为确定，不需要改；仅提示：这会让选择结果突破 32 条基础预算（文档已说明关键记录可突破预算）。

### 4. 大输入下的选择开销（信息性）
`select_record_indices` 对每条记录做 `value.to_string()` 用于去重，`reduce_records` 又对整个数组做一次序列化作为 Stash payload，峰值临时内存约为数组的 2-3 倍。`MAX_INPUT_BYTES` 为 64 MiB 且超时在压缩完成后才检查，最坏输入可能超过 timeout。当前可接受；后续可考虑用记录哈希替代整串存储来做去重。

### 5. 验证覆盖说明
本地未能运行 `make test-python-runtime`（本机仅 CPython 3.8，wheel 需要 cp311-abi3 工具链）。改动的两个 Python 测试（`test_runtime.py` 的 stash 初始化失败用例、`test_sdk.py` 的 post_tool + retrieve 往返）均为 Rust 侧已验证行为的薄封装，Rust 等价测试全部通过。

### 做得好的地方
- 确定性选择（BTreeSet 稳定顺序、众数 Shape 平局按字典序、中点分层采样）+ 33/32 阈值保证缩减必然至少省略 1 条；
- 每数组单一 Stash Entry、Marker 授权检索、Python SDK 端到端往返测试验证逐字节恢复；
- 无恢复路径的适配器保持无损优先、不做有损回退；干跑（压缩关闭）用 `InMemoryStore` 隔离，真实存储零写入；
- `StashLedger` 对孤儿写入的清理路径与本改动正确衔接（`commit` 删除无 Marker 的已创建行）；
- 文档（R1–R8 重新编号、示例 7、基准表）与基准阈值（55→35、52→49）随行为变化同步更新，且旧行为契约测试改用 `non_record_array_response`/字符串数组，意图清晰。

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

增量复审（@ 194384f）：24 个文件全部在 tokenless 范围内。本地复跑主工作区 `tokenless-compressors` / `tokenless-protocol` / `tokenless-runtime` 测试（162 通过）、L1（97 通过）、L2（42 通过），`cargo fmt --check` 与 `clippy -D warnings` 干净，与 Forrest-ly review 的验证结论一致。已对照三份既有 review 去重（Codex 的 omitted 计数 P2、ikunkun-sys 的 CLI 参考/Stash 契约文档 P2、Forrest-ly 的 4 项 🟡），不重复其已覆盖问题；新增 1 项非阻塞发现，无阻塞项。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

- 无

### 🟡 建议修改（不阻塞但推荐）

- [crates/tokenless-compressors/src/json.rs:455-492]（`select_numeric_outliers`，补充 Forrest-ly review 的 `has_error_signal` 条目）2σ 离群值兜底在低样本数字段上存在死区。方差用总体形式（除以 n），数学上 n 个样本中单点离群值可达的最大 z 值为 √(n-1)：当数值样本恰为 `NUMERIC_MIN_SAMPLES=5` 时上限为 2.0，而判定是严格大于（`(value - mean).abs() > threshold`），因此 5 样本字段的离群检测永远不会触发；n=6 起才可能生效，且要求其余样本足够均匀。实测验证：`[10,10,10,10,10000]` → z 恰为 2.0、不被选中；`[10,10,10,10,10,10000]` → z≈2.24、被选中。这与 Forrest-ly 指出的 `has_error_signal` 只匹配字符串状态的问题叠加：数值型状态码（如多数记录为 `"200"` 字符串、个别记录为 `"status": 500` 数值）本指望经 2σ 路径兜底，而此类错误稀少时（数值样本恰在 5~6 个）兜底恰好处在最弱或失效状态。因完整数组始终可从 Stash 取回，不影响可恢复性，仅影响 inline 保留集合。建议改用样本标准差（n-1 分母）、把 `NUMERIC_MIN_SAMPLES` 提到 ≥6，或改用中位数/MAD 这类稳健估计。

### 🟢 值得肯定

- 独立复核了 StashLedger 的 commit/rollback 语义：被拒绝候选（含 record reduction 的完整数组写入）只要标记未出现在最终输出即按 ownership 删除，不产生孤儿条目；`record_candidate_with_no_savings_rolls_back_its_single_write` 等测试本地复跑通过。
- Stash 不可用或写入失败时，33+ 项的 Object 数组不回退位置型截断、完整保留，杜绝了无可信恢复路径下的不可恢复丢失；选择逻辑（首尾 + 错误信号 + 结构异常 + 离群 + 去重稳定采样）确定且有专门测试钉住。

File: src/tokenless/crates/tokenless-compressors/src/json.rs (hunk 1 above)
Comment (Forrest-ly) [reply src/tokenless/crates/tokenless-compressors/src/json.rs:613]: Adopted. Updated the canonical EN/ZH user guides and the `compress-response` flag help to match the new contract (c2f82bea5):

- **CLI reference** (`cli-reference.md`, EN/ZH): the flag table now notes that `--truncate-arrays-at`/`--array-tail-preserve` do not apply to object record arrays, and a new paragraph explains that any array of at least 33 JSON objects is reduced to at most 32 sampled records plus a retrieval marker regardless of those flags, that the complete original array is written to the Stash as one entry for `tokenless retrieve`, and that `--no-stash` keeps every record of such arrays instead of reducing them.
- **Privacy guide** (`configuration-and-privacy.md`, EN/ZH): the Stash storage-boundary row and the "Sensitivity of Stash" section now state that record reduction stashes the complete original array before reduction, not only the omitted records.
- **User manual** (`user-manual.md`, EN/ZH): the Stash capability row updated with the same storage boundary.
- **CLI help** (`main.rs`): the `--truncate-arrays-at`, `--array-tail-preserve`, and `--no-stash` help strings no longer imply the old contract.

Verified against the built binary: 40 object records with `--truncate-arrays-at 1 --array-tail-preserve 0` emit 32 records plus the marker, `--no-stash` keeps all 40, and `tokenless retrieve` restores the complete original array byte-for-byte.

File: src/tokenless/crates/tokenless-compressors/src/json.rs [hunk 2]
```diff
@@ -422,6 +662,37 @@ impl<'a> Session<'a> {
...
+        let mut output = Vec::with_capacity(values.len());
+        for value in values {
+            self.push_if_kept(&mut output, value, depth);
+        }
+        Value::Array(output)
+    }
+
+    fn reduce_records(&mut self, values: &[Value], depth: usize) -> Option<Value> {
+        let selected = select_record_indices(values);
+        let omitted = values.len() - selected.len();
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply src/tokenless/crates/tokenless-compressors/src/json.rs:675]: Adopted. `reduce_records` now recounts `omitted` from the records actually emitted after recursive cleanup, so the marker and `records_omitted` stay consistent with the output (82f0e316b). The complete array is stashed either way, so retrieval is unaffected.

Added regression test `record_reduction_counts_selected_records_removed_by_cleanup` and verified via the CLI: a 40-record array whose first record collapses to `{}` during cleanup now reports `9 of 40 records omitted` with 31 records emitted, and `tokenless retrieve` round-trips the full original array exactly.

**Review by Forrest-ly — APPROVED** [review]
Re-review of the 2 new commits since @194384f (incremental diff: 9 files, +62/−15). **APPROVE** — both P2 issues are fixed correctly, with tests; one non-blocking doc wording nit below.

## `82f0e31` — records_omitted recount (Codex bot P2 / Code Reviewer suggestion #2)

Verified correct in `src/tokenless/crates/tokenless-compressors/src/json.rs`:

- The recount `values.len() - output.len()` is computed at the point where `output` holds exactly the emitted selected records (the marker is pushed afterwards), so the marker and `records_omitted` are now consistent with the visible output. No underflow possible (`output.len() ≤ selected.len() ≤ values.len()`).
- The complete original array is serialized and stashed **before** the selection loop (`serde_json::to_string(values)`), so cleanup-collapsed records don't affect retrieval — matches the commit message.
- Edge case where cleanup collapses *all* selected records degrades gracefully: output becomes marker-only, `omitted == values.len()`.
- New test `record_reduction_counts_selected_records_removed_by_cleanup` covers the scenario end-to-end: kept count, operation list, metrics, exact marker text, recoverability, and byte-exact stash retrieval.

## `c2f82be` — docs contract alignment (ikunkun-sys P2)

Cross-checked every claim against the code:

- "≥ 33 objects → reduced" ✓ (`RECORD_MIN_ITEMS = 33`, `is_record_array`)
- "first 4 / last 4 + error & anomaly records + stable sample" ✓ (`select_record_indices`)
- "regardless of `--truncate-arrays-at` / `--array-tail-preserve`" ✓ (record-array path runs before the head/tail window logic in `compress_array`)
- "`--no-stash` keeps every record instead of reducing" ✓ (`reduce_records` → `None` without stash falls back to `compress_all_array_items`, never positional truncation; covered by `record_array_without_stash_does_not_fall_back_to_positional_truncation` and `record_stash_failure_keeps_every_record`)
- EN/ZH updates are consistent, and the CLI help text in `main.rs` matches.

**Non-blocking nit:** EN `cli-reference.md` says "reduced to **at most 32 records**" (ZH: "**最多** 32 条记录"), but critical records can legitimately exceed the budget — the code explicitly allows it (`critical_records_can_exceed_the_normal_budget` asserts `selected.len() > RECORD_MAX_ITEMS`) and the PR description itself calls 32 a "**base** budget". Since this commit exists to align docs with the contract, suggest softening to "typically 32 records / a base budget of 32 records", noting error/anomaly records may exceed it.

## Carry-over (still non-blocking, no action required in this PR)

`select_numeric_outliers` 2σ dead zone from the earlier incremental review: population variance + 2σ can still miss outliers on low-cardinality numeric fields near `NUMERIC_MIN_SAMPLES = 5`. Options if revisited later: sample stddev (N−1), higher min samples, or MAD.

## Local verification @c2f82be

- `cargo fmt --all -- --check` — clean
- `cargo clippy --workspace --all-targets --locked -- -D warnings` — clean
- `cargo test --workspace --locked` — all suites pass, 0 failures
- New regression test run individually — pass

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

增量审查本轮两个新提交（`82f0e31` omitted 计数修正、`c2f82be` 用户文档对齐），增量 diff 9 个文件 +62/−15。计数修正逻辑正确且回归测试充分；文档对齐基本完整，仅发现一处被修复提交遗漏的旧表述（补充 ikunkun-sys review 的 P2 线索）。本地验证：`cargo test -p tokenless-compressors`（80 passed）与 `cargo test -p tokenless-cli`（278 passed）均通过，新增回归测试单独跑过。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

- 无

### 🟡 建议修改（不阻塞但推荐）

- [docs/user-guide/en/token-saving/tokenless/cli-reference.md:244、docs/user-guide/zh/token-saving/tokenless/cli-reference.md:237]（补充 ikunkun-sys 的 P2 review）c2f82be 修复了 ikunkun-sys 指出的"canonical guides 仍描述旧 array/Stash 契约"问题，但同一文件末尾的 Stash 范围句被遗漏：EN 仍写 "Stash applies **only** to truncation of strings, the dropped middle segment of truncated arrays, and deep subtrees"，ZH 仍写 "Stash **只**作用于字符串、截断数组中被丢弃的中间段和深层子树截断"。而记录缩减会把完整原始数组写入 Stash——同文件上方新增段落（EN 225 行）和本提交更新后的 configuration-and-privacy.md 数据表都明确了这一点，"only/只" 使该句与文件自身及隐私文档矛盾，用户仍可能据旧句低估 `stash.db` 的敏感数据面。建议把记录缩减的完整数组并入该枚举（或去掉 "only/只"），EN/ZH 同步修改。

### 🟢 值得肯定

- `82f0e31` 的修正位置和语义都正确：重新计数发生在 marker 入栈之前，`output` 此时恰好包含实际输出的被选记录；`select_record_indices` 返回 `BTreeSet<usize>` 且 `push_if_kept` 每条至多推入一项，`output.len() ≤ selected.len() ≤ values.len()`，无下溢风险；完整数组在选择循环之前序列化入 Stash，cleanup 折叠不影响取回；全部被折叠时退化为仅 marker 输出且 `omitted == values.len()`，语义仍自洽。新测试 `record_reduction_counts_selected_records_removed_by_cleanup` 端到端钉住了保留数、操作序列、`records_omitted`、精确 marker 文本、可恢复性和 stash 逐字节取回。
- `c2f82be` 的文档主张逐条对照代码核实无误：≥33 个 Object 触发（`RECORD_MIN_ITEMS = 33` / `is_record_array`）、首尾各 4 条 + 错误/异常信号 + 稳定采样、不受 `--truncate-arrays-at`/`--array-tail-preserve` 影响（记录数组分支先于头尾窗口逻辑）、`--no-stash` 时 `reduce_records` 返回 None 后走 `compress_all_array_items` 完整保留而非位置型截断、stash 条目为缩减前完整数组；EN/ZH 表述一致，main.rs 三个 flag 的 help 文本同步更新且与实际行为相符。

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

对新 head `9f4d55f` 做增量审查（基线为此前已批准的 `c2f82be` + 用户已批准的增量）。autosquash 未引入任何内容偏差：`c2f82be → 9f4d55f` 的树差异仅 4 个文件（EN/ZH cli-reference.md、main.rs、cli_integration.rs），恰好等于本轮跟进修复内容，此前已审查的全部改动逐字节保留。三项跟进修复（post_tool dry-run 挂载配置的 Stash、CLI reference 32 条基础预算措辞、Stash 数据边界枚举）均已正确落实。本地验证：`cargo fmt --check` 干净，`cargo clippy --workspace --all-targets --locked -- -D warnings` 干净，`cargo test --workspace --locked` 707 passed / 3 ignored 全绿。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

- 无

### 🟡 建议修改（不阻塞但推荐）

- [docs/user-guide/zh/token-saving/tokenless/cli-reference.md:237] 中文术语不一致：Stash 范围句使用未翻译的 "Stash 会保存 Record Reduction 使用的完整原始数组"，而同文件其余位置（207、218 行）及 configuration-and-privacy.md:127、user-manual.md:49 均用 "记录缩减"。建议统一为 "记录缩减"（如 "Stash 会保存记录缩减使用的完整原始数组……"），一行改动，不阻塞。

### 🟢 值得肯定

- dry-run Stash 修复的设计正确且分层完备：`main.rs` 中 PostTool 的 `needs_stash` 解除对 `compression_on` 的依赖后，`pipeline.rs:88-101` 的 `dry_run_store`（InMemoryStore）机制才真正经 CLI 路径可达；`decide()` 保证 `dry_run` 时绝不返回 Apply（`arbitration.rs:44-46`），DryRun/Reject 均 `ledger.rollback`，且 `persistent_store_attached = compression_enabled && attached_store.is_some()` 使干跑时 `stash_writes/stash_errors/stash_size` 报告均为 None，持久 store 零写入有结构性保证。
- 测试覆盖双层互补：既有管线级 `record_reduction_dry_run_uses_only_a_temporary_store`（钉住操作序列为 RecordReduction、对持久 CountingStore 的 stash/delete 调用均为 0）加上新增 CLI 级 `compress_post_tool_dry_run_previews_record_reduction_without_stash_writes`（TOKENLESS_COMPRESSION_ENABLED=0 下 200 条记录：disposition 为 dry_run、输出逐字节等于原文、after_tokens < before_tokens、真实 SQLite stash.db 条目数为 0），正好落实 08:49 UTC 评审的可执行项。
- autosquash 忠实：`git diff c2f82be 9f4d55f` 仅上述 4 个文件 +52/−13，无其他内容偏差；全工作区 707 测试通过（较上轮 705 恰好多 2 个新增测试）。
- 文档措辞与代码契约一致："reduced against a base budget of 32 records … critical records can exceed the base budget"（ZH："按 32 条记录的基础预算进行缩减……关键记录可以突破基础预算"）与 `critical_records_can_exceed_the_normal_budget` 断言的 `selected.len() > RECORD_MAX_ITEMS` 行为相符，并补上了数值异常（numeric outliers）选择信号；同时落实了 10:40 UTC 评审指出的 Stash 范围句——EN/ZH 均去掉 "only/只" 并把记录缩减的完整原始数组并入枚举，与 configuration-and-privacy.md 的数据边界不再矛盾。
- PR 描述已精确化 15% 规则（"savings of at least 15% immediately select that candidate and skip bounded work, while smaller savings continue to bounded generation and final arbitration"），与 `saves_at_least_percent` 仅作提前接受阈值的实现一致。

### 信息性说明（无需改动）

- 行为面变化：post_tool 干跑（请求声明 retrieval_available 时）现在会打开配置的 stash.db——首次干跑会创建一个空的 0600 数据库文件，若数据库无法打开则与压缩开启路径一致地退出 1。这与"干跑验证配置的恢复路径"的设计意图相符，且失败面与真实运行一致（0/1 双跑对照不会因此失真）；隐私文档 "Dry-run does not create Stash entries"（configuration-and-privacy.md:166、measuring-savings.md:204）仍然准确。若后续想更进一步，可考虑为"干跑 + stash 库无法打开 → 退出 1"补一个 CLI 级回归测试，非必需。

**Review by Forrest-ly — APPROVED** [review]
## 总体评价

Re-review 针对 11:04:52Z 的 squash force-push（c2f82be → 9f4d55f）。已核实：`git diff c2f82bea..9f4d55fb` 恰为 4 文件 +52/−13，与 squash 前已审查内容完全一致，未丢失/篡改任何已审查改动。增量包含三部分内容：① 采纳上一轮非阻塞意见 #1 的文档措辞软化（EN/ZH cli-reference.md）；② main.rs 的 PostTool dry-run 行为变更（此前各轮均未覆盖）；③ 对应的新增集成测试。

本地验证（9f4d55fb）：`cargo fmt --check` 通过；`cargo clippy --workspace --all-targets -- -D warnings` 零告警；`cargo test --workspace` 707 passed / 3 ignored / 0 failed（20 suites），其中新增测试 `compress_post_tool_dry_run_previews_record_reduction_without_stash_writes` 单独跑过通过。

## 审查结论

approve

## 详细意见

### 🔴 必须修改（阻塞合并）

- 无

### 🟡 建议修改（不阻塞但推荐）

- 无（两条观察见 🟢 之后的"备注"）

### 🟢 值得肯定 / 验证通过项

1. **squash 完整性**：增量与描述完全吻合（docs EN/ZH cli-reference.md、main.rs、cli_integration.rs，+52/−13），无隐藏改动。
2. **main.rs PostTool dry-run 语义**（`src/tokenless/crates/tokenless-cli/src/main.rs:514-520`）：PostTool 的 `needs_stash` 不再叠加 `compression_on`，dry-run 也会打开配置的 store。与 `open_stash_store_or_err_with` 的硬失败语义（失败 → exit 1）组合后，dry-run 与真实压缩运行在 store 不可用时行为一致——预览如实反映真实运行的可用性，符合"verify that recovery is available"的注释意图。
3. **无持久化副作用泄漏**：`tokenless-runtime/src/post_tool/pipeline.rs:88-100` 在 `!compression_enabled && stash_enabled && retrieval_available && stash_store.is_some()` 时用 `InMemoryStore` 替代持久 store 做候选生成；`Verdict::DryRun` 分支对内存 store 执行 ledger.rollback 并原样输出原文、返回空 stash_keys。持久 store 在 dry-run 下不会收到任何 Stash 条目——与注释及文档一致。
4. **新集成测试充分**：200 条记录（远超 33 阈值）+ `TOKENLESS_COMPRESSION_ENABLED=0` + stats/SLS 关闭，断言 disposition=dry_run、output==原文、after_tokens<before_tokens（预览了缩减）、持久 `stash.db` 条目数为 0，精确覆盖新行为。
5. **文档措辞**（非阻塞意见 #1 已完整采纳）：EN/ZH 均改为"按 32 条记录的基础预算缩减…关键记录可突破基础预算"，并补充数值异常记录与 Stash 适用范围含完整原始数组；全库已无 "at most 32/最多 32" 残留，`stash-reversible-compression.md` 亦一致。
6. **沿用前轮结论**：`select_numeric_outliers` 2σ 兜底在小样本数字段上的死区，维持"无需本 PR 处理"。

### 备注（非阻塞观察）

- dry-run 失败模式变化：PostTool 且 `retrieval_available=true` 时，若 stash db 无法打开/创建，dry-run 现在会 exit 1（此前完全跳过 store）。这是有意语义（与真实运行对齐，PR 描述已声明公开行为变更），且 `ensure_db_dir` 会自动建目录、空库以 0600 权限创建，风险低；仅提示 adapter 侧 dry-run 基线将走与真实运行相同的 fail-open 路径。
- `configuration-and-privacy.md` 的 "Dry-run does not create Stash entries" 仍然严格成立（条目 ≠ 空库文件），但 dry-run 现可能创建一个空 `stash.db`；如需绝对零落盘，现有文档建议（关 stats/SLS）之外可考虑补一句。不阻塞。

**Review by ikunkun-sys — APPROVED** [review]
Re-reviewed the updated head. The two earlier P2 issues are resolved: the omitted count is recomputed from records actually emitted after cleanup, and the canonical EN/ZH guides now document the record-reduction flags and complete-array Stash boundary. Rust 1.89 fmt, all-target clippy, workspace tests, rustdoc, docs lint/link checks, the synthetic merge, and production CLI reduction/TOON/retrieval/dry-run/no-stash probes passed. No new P0/P1/P2 findings.

**Review by Forrest-ly — APPROVED** [review]
Code review 通过（re-review @ d0aee2f）。新 head 相对此前已审版本完整复核，结论 APPROVE。

## 复核范围
单 commit（autosquash）d0aee2fa，31 文件 +1535/−209。核心改动：
- `tokenless-compressors/src/json.rs`：无损候选优先的两段式选择（完整数据 Compact JSON/TOON 达到 15% token 节省即跳过 bounded 处理），以及 ≥33 项全 Object 数组的 Record Reduction（首 4 + 尾 4 + 错误信号 + 结构异常 + 2σ 数值异常，去重后等距采样至 32 条基础预算，关键记录可突破预算）。
- `tokenless-runtime`：PostTool dry-run 用 InMemoryStore 生成候选（持久 store 零写入），`allow_unrecoverable` 接入 `require_reversibility`；stats 新增 `json_record_reduction` 标签。
- Protocol/SDK：新增 `AppliedOperation::JsonRecordReduction`（wire 名 `json_record_reduction`，Rust/Python 一致）。

## 已验证
- 设计正确性：完整数组先落 Stash 再输出缩减结果与 retrieve marker；stash 缺失或写入失败时保留全部记录、不回退位置型截断（`record_array_without_stash_does_not_fall_back_to_positional_truncation` / `record_stash_failure_keeps_every_record` 固定）；`recoverability` 将 record reduction 计入非无损，无 stash 场景不会产生虚假 Retrievable。
- dry-run：CLI 打开配置 store 做可用性校验（打不开 exit 1，有意语义），候选生成走内存 store，`stash_writes/stash_errors/stash_size` 仅 compression 开启时上报；`compress_post_tool_dry_run_previews_record_reduction_without_stash_writes` 验证 stash.db 0 条目。
- 契约 golden：`bash_array_truncation/cosh-ng.json` 由空 envelope 变为完整输出——已用 hermetic hook runner 复现：hook 声明 `retrieval_available=false`，旧行为下 bounded 候选 Unrecoverable 被拒（输出 `{}`）；新无损优先路径选中超 15% 门槛的 TOON 表示（lossless applied），属于预期且有利的行为变化；其余 4 个 agent golden 不变（replace_with_text=false 不允许 TOON）也一致。
- 测试：主工作区 `cargo test` 707 passed / 0 failed / 3 ignored；l1-compressor 97 passed、l2-module 42 passed（含新增 `record_reduction_fixture_preserves_the_middle_anomaly`、调整后的 35%/49% 节省阈值）；`cargo fmt`、`cargo clippy --workspace --all-targets` 全绿；`fixtures/record_reduction.json` 与 gen_fixtures.py 逐字节一致。
- 文档：response-compression / stash-reversible-compression / CLI reference / user-manual（EN+ZH）与实现一致（8 条规则、33 项门槛、32 条基础预算、--no-stash 跳过缩减）。

## 非阻塞备注（沿用上轮结论）
1. dry-run 会打开（并可能创建空）持久 stash.db——有意的可用性校验，代码注释已说明。
2. `select_numeric_outliers` 2σ 在 5–10 个小样本下的选择倾向为启发式，确定性且有界。
3. 33–36 项数组可能仅省略 1–4 条即写入完整数组 stash，净收益小但可恢复、标记一致。
4. 无损候选未达 15% 时会执行两次压缩遍历——设计取舍，已知。

本地环境备注：l1_adversarial_schema 的 `very_deep_nested_schema_does_not_overflow` 在默认 2MB 测试线程栈下溢出，已确认在父提交 6712b3d6 上同样出现且 `RUST_MIN_STACK=16MB` 下通过，与本 PR 无关。
